### PR TITLE
API Updates

### PR DESCRIPTION
**Pull request by Entrox-Licher**
Adding support across the codebase to utilize new and existing APIs following the API pattern.

The code here is reliant on the ChaosBagApi implemented by the Campaign Importer/Exporter, so will need that to be merged in first.

### CODE REVIEW

File: src/core/GameKeyHandler.ttslua [hunk 1]
```diff
@@ -98,14 +99,18 @@ function takeClueFromLocation(playerColor, hoveredObject)
...
+  local matColor = playmatApi.getMatColor(playerColor)
+  local pos = nil
+  if clickableClues then
+    pos = {x = 0.49, y = 2.66, z = 0.00}
+    playmatApi.updateClueClicker(playerColor, playmatApi.getClueCount(clickableClues, playerColor) + 1)
+  else
+    pos = playmatApi.transformLocalPosition({x = -1.12, y = 0.05, z = 0.7}, matColor)
   end
+  
+  local rot = playmatApi.returnRotation(matColor)
```
Comment (Chr1Z93) [inline src/core/GameKeyHandler.ttslua:113]: I think you need to add a replacement effect to the physical move of the clue token - see LOC 116-121 in your code (like moving it to a trash can instead).

File: src/core/GameKeyHandler.ttslua [hunk 2]
```diff
@@ -98,14 +99,18 @@ function takeClueFromLocation(playerColor, hoveredObject)
...
 
   -- check if found clue is a stack or single token
```
Comment (Chr1Z93) [inline src/core/GameKeyHandler.ttslua:115]: The following lines need handling for clickable clues

File: src/accessories/CleanUpHelper.ttslua [hunk 3]
```diff
@@ -8,6 +8,7 @@ local soundCubeApi              = require("core/SoundCubeApi")
 local playmatApi                = require("playermat/PlaymatApi")
 local blessCurseManagerApi      = require("chaosbag/BlessCurseManagerApi")
 local chaosBagApi               = require("chaosbag/ChaosBagApi")
+local playAreaAPI               = require("core/PlayAreaApi")
```
Comment (Chr1Z93) [inline src/accessories/CleanUpHelper.ttslua:11]: "API" should be "Api" for consistent naming

File: src/core/PlayAreaApi.ttslua [hunk 4]
```diff
@@ -83,5 +84,30 @@ do
...
+    return getObjectFromGUID(PLAY_AREA_GUID).getCustomObject()
+  end
+
+  PlayAreaApi.setCustomObject = function(customInfo)
+    return getObjectFromGUID(PLAY_AREA_GUID).setCustomObject(customInfo)
+  end
+
+  PlayAreaApi.reload = function()
+    return getObjectFromGUID(PLAY_AREA_GUID).reload()
+  end
```
Comment (Chr1Z93) [inline src/core/PlayAreaApi.ttslua:108]: Why do we need these as functions in the API?

File: src/core/GameKeyHandler.ttslua (hunk 2 above)
Comment (Chr1Z93) [reply src/core/GameKeyHandler.ttslua:115]: True, I missed that you already implemented that

File: src/core/PlayAreaApi.ttslua (hunk 4 above)
Comment (Chr1Z93) [reply src/core/PlayAreaApi.ttslua:108]: Usually we add an Api if we need to access resources (variables or functions) of another object. So for this case I'd say it would be better if you just add a new function that takes the new image as parameter instead of 3 functions that are useless on their own.

File: src/core/GameKeyHandler.ttslua [hunk 5]
```diff
@@ -98,14 +99,18 @@ function takeClueFromLocation(playerColor, hoveredObject)
...
   end
 
+  local clickableClues = optionPanelApi.getOptions()["useClueClickers"]
   local playerName = Player[playerColor].steam_name
-  local matColor = playmatApi.getMatColor(playerColor) 
-  local pos = playmatApi.transformLocalPosition({x = -1.12, y = 0.05, z = 0.7}, matColor)
-  local rot = playmatApi.returnRotation(matColor)
-
-  if cardName == "" or cardName == nil then
-    cardName = "nameless card"
```
Comment (Chr1Z93) [inline src/core/GameKeyHandler.ttslua:107]: Why was this removed?

File: src/core/PlayArea.ttslua [hunk 6]
```diff
@@ -504,20 +504,6 @@ function shiftContents(playerColor, direction)
...
-  return investigatorCounter.getVar("val")
-end
-
--- Updates the current value of the investigator counter from the playmat
----@param count Number of investigators to set on the counter
-function setInvestigatorCount(count)
-  local investigatorCounter = getObjectFromGUID("f182ee")
-  return investigatorCounter.call("updateVal", count)
-end
-
```
Comment (Chr1Z93) [inline src/core/PlayArea.ttslua:520]: Can you confirm that you checked whether this is referenced by any other scripts?

File: src/core/PlayAreaApi.ttslua [hunk 7]
```diff
@@ -2,19 +2,20 @@ do
   local PlayAreaApi = { }
 
   local PLAY_AREA_GUID = "721ba2"
+  local INVESTIGATOR_COUNTER_GUID = "f182ee"
 
   local IMAGE_SWAPPER = "b7b45b"
```
Comment (Chr1Z93) [inline src/core/PlayAreaApi.ttslua:7]: While you are at it, this variable should get a `_GUID` suffix too

File: src/core/PlayAreaSelector.ttslua [hunk 8]
```diff
@@ -70,8 +71,7 @@ function none() end
 
 -- main function (can be called by other objects)
 function updateSurface(newURL)
-    local playArea = getObjectFromGUID("721ba2")
-    local customInfo = playArea.getCustomObject()
+    local customInfo = playAreaAPI.getCustomObject()
```
Comment (Chr1Z93) [inline src/core/PlayAreaSelector.ttslua:74]: Instead of adding a single-use function to the Api, let's add the "updateSurface" functionality as an Api function

File: src/playermat/Playmat.ttslua [hunk 9]
```diff
@@ -772,7 +774,7 @@ end
 ---------------------------------------------------------
 
 function drawChaosTokenButton(_, _, isRightClick)
-  Global.call("drawChaosToken", {self, DRAWN_CHAOS_TOKEN_OFFSET, isRightClick})
+  chaosBagApi.drawChaosToken({self, DRAWN_CHAOS_TOKEN_OFFSET, isRightClick})
```
Comment (Chr1Z93) [inline src/playermat/Playmat.ttslua:777]: Advantage of the Api files is that we don't have to restrict us to a single parameter (and thus make multiple ones into a single table). So'd split these into three separate parameters.

File: src/playercards/AllCardsBagApi.ttslua
Comment (Chr1Z93) [inline src/playercards/AllCardsBagApi.ttslua:1]: While you can only pass a single parameter to a function call on another object, we don't have that limitation in the Api's itself. Thus I think it would be better if we split those up. Example:
```
before:
AllCardsBagApi.getCardsByName = function(params)
  return getObjectFromGUID(ALL_CARDS_BAG_GUID).call("getCardsByName", params)
end

after:
AllCardsBagApi.getCardsByName = function(name, exact)
  return getObjectFromGUID(ALL_CARDS_BAG_GUID).call("getCardsByName", {name = name, exact = exact)
end
```
Of course this also requires a change to the places that use the Api function.

File: src/arkhamdb/DeckImporterUi.ttslua [hunk 10]
```diff
@@ -218,8 +220,7 @@ function loadDecks()
   -- Method in DeckImporterMain, visible due to inclusion
 
   -- TODO: Make this use the configuration ID for the all cards bag
```
Comment (Chr1Z93) [inline src/arkhamdb/DeckImporterUi.ttslua:222]: can remove the "TODO" line at this point

**Review by Chr1Z93 — CHANGES_REQUESTED** [review]
See separate comments

File: src/core/GameKeyHandler.ttslua (hunk 5 above)
Comment (Entrox-Licher) [reply src/core/GameKeyHandler.ttslua:107]: Because it conflicts with the if statement a few lines down (LOC 123-127).  Of the two options, I thought the later one made more sense, since stating you received a clue from a 'nameless card' is not very useful, so you may as well just say nothing about the card at all.  I can switch it back if preferred though, but I still think we should have only one or the other for code clarity.

File: src/core/PlayArea.ttslua (hunk 6 above)
Comment (Entrox-Licher) [reply src/core/PlayArea.ttslua:520]: Yep!  It only existed for the API, so I just put it in there for clarity

File: src/playermat/Playmat.ttslua (hunk 9 above)
Comment (Entrox-Licher) [reply src/playermat/Playmat.ttslua:777]: Yep, makes perfect sense.  I was trying to just make them behave the same as the original function for consistency, but I agree it's just much more usable to split them all up

File: src/playercards/AllCardsBagApi.ttslua
Comment (Entrox-Licher) [reply src/playercards/AllCardsBagApi.ttslua:1]: Agreed!  My only concern is they become slightly inconsistent (i.e. when used internally they are referenced differently to when they are used externally, since the internal function still needs to only take a single parameter in order to make the call possible in the API at all).  That being said, since these functions are used in so many other places, and the difference is relatively minor, I don't have an issue with it

Comment (Entrox-Licher) [reply src/playercards/AllCardsBagApi.ttslua:1]: Note: latest commit has implemented this change, however due to TTS being a little ridiculous with their `Attempt to perform operations with resources owned by different scripts` error, the `getCardById` function still needs to be a params table, even though it's only one parameter; it will trigger the aforementioned error otherwise, for reasons known only to The Creators

File: src/arkhamdb/ArkhamDb.ttslua [hunk 11]
```diff
@@ -166,18 +166,20 @@ do
...
-      local weaknessId = allCardsBagApi.getRandomWeaknessId()
-      slots[weaknessId] = 1
-      slots[RANDOM_WEAKNESS_ID] = nil
-      internal.maybePrint("Random basic weakness added to deck", playerColor)
+    if randomWeaknessAmount ~= 0 then
+      for i=1, randomWeaknessAmount do
+        local weaknessId = allCardsBagApi.getRandomWeaknessId()
+        slots[weaknessId] = 1
+        slots[RANDOM_WEAKNESS_ID] = nil
+      end
```
Comment (Chr1Z93) [inline src/arkhamdb/ArkhamDb.ttslua:181]: Now that I read this again, I think this will result in too few weaknesses if the random weakness rolls the same one multiple times 😅
We're gonna need to check if the weaknessId was already added to the deck (`if slots[weaknessId] == 1`) and get a new one if it is already in there

Comment (Chr1Z93) [reply src/arkhamdb/ArkhamDb.ttslua:181]: Hm and I'm not certain whether this will take the weakness count into consideration when importing multiple decks. Theoretically, it should since they are load one by one, but I am not sure if there is enough delay for the allcardsbag to pick up on the weaknesses in other peoples decks (and also doesn't work properly if a unique weakness is in a deck that will be loaded later)

Comment (Chr1Z93) [reply src/arkhamdb/ArkhamDb.ttslua:181]: I'm probably going to target this in a separate PR since this one is already huge

**Review by Chr1Z93 — CHANGES_REQUESTED** [review]

**Review by Chr1Z93 — APPROVED** [review]
